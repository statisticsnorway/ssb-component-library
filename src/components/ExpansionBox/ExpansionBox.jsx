import React, { useEffect, useRef, useState } from 'react'
import PropTypes from 'prop-types'
import { ChevronDown, ChevronUp } from 'react-feather'

const ExpansionBox = ({
  className = '',
  header = '',
  icon,
  onToggle = () => {},
  openByDefault = false,
  sneakPeek,
  text = '',
}) => {
  const [isOpen, toggleOpen] = useState(openByDefault)
  const firstUpdate = useRef(true)
  useEffect(() => {
    if (firstUpdate.current) {
      firstUpdate.current = false
      return
    }
    onToggle(isOpen)
  }, [isOpen])
  return (
    <div className={`ssb-expansion-box${className ? ` ${className}` : ''}${sneakPeek ? ` sneak-peek` : ''}`}>
      <button
        className={`header ${isOpen ? 'open' : 'closed'}`}
        aria-expanded={isOpen ? 'true' : 'false'}
        onClick={() => toggleOpen(!isOpen)}
      >
        <span className='button-grid'>
          {icon && <div className='icon'>{icon}</div>}
          <span className='header-text'>{header}</span>
          <div className='icon-wrapper'>
            {!isOpen && <ChevronDown className='expand-icon' size={24} />}
            {isOpen && <ChevronUp className='expand-icon' size={24} />}
          </div>
        </span>
      </button>
      <div className={`content ${isOpen ? 'open' : 'closed'}`}>{text}</div>
    </div>
  )
}

ExpansionBox.propTypes = {
  className: PropTypes.string,
  header: PropTypes.string.isRequired,
  icon: PropTypes.node,
  onToggle: PropTypes.func,
  openByDefault: PropTypes.bool,
  sneakPeek: PropTypes.bool,
  text: PropTypes.oneOfType([PropTypes.string, PropTypes.node]).isRequired,
}

export default ExpansionBox
