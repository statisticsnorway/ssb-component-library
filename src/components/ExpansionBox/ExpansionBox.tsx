import React, { useState, ReactNode } from 'react'
import { ChevronDown, ChevronUp } from 'react-feather'

export interface ExpansionBoxProps {
  className?: string
  header: string
  icon?: ReactNode
  openByDefault?: boolean
  sneakPeek?: boolean
  text: string | ReactNode
}
const ExpansionBox: React.FC<ExpansionBoxProps> = ({
  className = '',
  header = '',
  icon,
  openByDefault = false,
  sneakPeek,
  text = '',
}) => {
  const [isOpen, toggleOpen] = useState(openByDefault)

  return (
    <div
      className={`ssb-expansion-box${className ? ` ${className}` : ''}${isOpen ? ' open' : ''}${sneakPeek ? ` sneak-peek` : ''}`}
    >
      <button className='header' aria-expanded={isOpen ? 'true' : 'false'} onClick={() => toggleOpen(!isOpen)}>
        {icon && <div className='icon'>{icon}</div>}
        <span className='header-text'>{header}</span>
        <div className='icon-wrapper'>
          {!isOpen && <ChevronDown className='expand-icon' size={24} />}
          {isOpen && <ChevronUp className='expand-icon' size={24} />}
        </div>
      </button>
      <div className={`content ${!isOpen ? 'closed' : ''}`}>{text}</div>
    </div>
  )
}

export default ExpansionBox
