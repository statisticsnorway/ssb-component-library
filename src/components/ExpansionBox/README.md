# ExpansionBox

> Wrapper for expansionBox

### Usage

#### HTML

```html Default
<div class='ssb-expansion-box'>
  <button aria-expanded='false' class='header'>
    <span class='header-text'>This is a ExpansionBox title</span>
    <div class='icon-wrapper'>
      <i class="expand-icon">{feather.chevronDown 20px}</i>
    </div>
  </button>
  <div class='content closed'>This is paragraph text</div>
</div>
```

```html SneakPeek and aiIcon
<div class="ssb-expansion-box sneak-peek">
   <button class="header" aria-expanded="false">
      <div class="icon">
        <i>{SparklesIcon 32px}</i>
      </div>
      <span class="header-text">ExpansionBox with icon and sneakpeek content</span>
      <div class="icon-wrapper">
         <i class="expand-icon">{feather.chevronDown 20px}</i>
      </div>
   </button>
   <div class="content closed">
      {content}
   </div>
</div>
```

#### React

```jsx harmony
<ExpansionBox header='Default expansionBox' text='This is paragraph text which explains the expansionBox' />

<ExpansionBox header='ExpansionBox open by default' text='This is paragraph text which explains the expansionBox' openByDefault />

<ExpansionBox header='ExpansionBox with AI icon and sneakpeek text' text='This is paragraph text which explains the expansionBox' aiIcon sneakPeek />
```

Available props:

| Name          | Type              | Description                                                    |
| ------------- | ----------------- | -------------------------------------------------------------  |
| aiIcon        | bool              | Show a AI icon to the left of header                           |
| className     | string            | Optional container class                                       |
| header        | string            | Header text                                                    |
| sneakPeek     | bool              | Show a sneakpeek of the text                                   |
| openByDefault | bool              | Open when rendered. Defaults to false                          |
| text          | string or element | Content in expansionBox                                        |


