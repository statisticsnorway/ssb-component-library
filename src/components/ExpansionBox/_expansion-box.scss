.ssb-expansion-box {
  border: 1px solid $ssb-green-4;
  border-radius: 8px;
  display: grid;
  width: 100%;

  &:has(.header:hover) {
    box-shadow: 0 0 0 1px $ssb-green-4;
  }

  .header {
    @include reset-button;
    padding: 1.7rem 2.5rem;
    align-items: center;
    cursor: pointer;
    display: flex;

    &:hover {
      .icon-wrapper {
        &::before {
          height: 3rem;
          width: 3rem;
        }

        .expand-icon {
          color: $ssb-white;
        }
      }
    }

    .header-text {
      @include roboto;
      color: $ssb-dark-6;
      font-size: 1.25rem;
      font-weight: bold;
      margin-right: 1rem;
    }

    .icon-wrapper {
      display: inline-block;
      margin-left: auto;
      position: relative;
      height: 1.5rem;
      width: 1.5rem;

      &::before {
        content: '';
        position: absolute;
        top: 50%;
        left: 50%;
        width: 0;
        height: 0;
        background-color: $ssb-green-4;
        border-radius: 50%;
        transform: translate(-50%, -50%);
        z-index: 0;
      }

      .expand-icon {
        color: $ssb-green-4;
        position: relative;
        z-index: 1;
        height: 1.5rem;
        width: 1.5rem;
      }
    }

    .icon {
      align-items: center;
      display: inline-flex;
      margin-right: 1rem;

      svg {
        height: 2rem;
        width: 2rem;
      }
    }

    &:focus-visible {
      outline: none;
    }

    @media #{$mobile} {
      padding: 1.25rem;
    }
  }

  .content {
    @include open-sans;
    color: $ssb-dark-6;
    font-size: 1rem;
    line-height: 1.75rem;
    overflow: hidden;
    padding: 0 4rem 0 2.5rem;

    &.closed {
      max-height: 0;
      opacity: 0;
      transition:
        max-height 0.5s,
        opacity 0.5s;
    }
  }

  &.open {
    &:focus-within {
      outline: none;
    }

    .header:focus-visible {
      @include focus-ring;
      border-radius: 8px;
    }

    .content {
      max-height: 1000px;
      opacity: 1;
      overflow-y: auto;
      padding: 0 4rem 2rem 2.5rem;
      transition: max-height 0.5s;

      @media #{$mobile} {
        padding: 0 1.25rem 1.25rem;
      }
    }
  }

  &:focus-within {
    @include focus-ring;
    border-radius: 8px;
  }

  &.sneak-peek {
    &:not(&.open) {
      position: relative;

      .header::before {
        content: '';
        position: absolute;
        inset: 0;
        z-index: 1;
      }
    }

    .header {
      padding-bottom: 0.5rem;
      transition: padding-bottom 0.5s;
    }

    &.open .header {
      padding-bottom: 2.5rem;
      transition: padding-bottom 0.5s;
    }

    .content.closed {
      position: relative;
      max-height: 5.5rem;
      opacity: 1;
      margin-bottom: 1.5rem;

      &::after {
        content: '';
        position: absolute;
        inset: 0;
        background-image: linear-gradient(to bottom, rgba(255 255 255 / 0%) 50%, rgba(255 255 255) 100%);
      }
    }
  }
}
