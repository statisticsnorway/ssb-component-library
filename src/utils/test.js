import fs from 'fs';
import path from 'path';
import { render as rtlRender } from '@testing-library/react';

export * from '@testing-library/react';
export * from '@testing-library/jest-dom';

const styles = fs.readFileSync(path.relative(process.cwd(), './lib/bundle.css'), 'utf8');

export function render(ui, options) {
	const view = rtlRender(ui, options);

	const styleElement = document.createElement('style');
	styleElement.innerHTML = styles;
	document.body.appendChild(styleElement);

	return view;
}
