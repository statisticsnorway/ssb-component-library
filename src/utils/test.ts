import fs from 'fs'
import path from 'path'
import { RenderOptions, RenderResult, render as rtlRender } from '@testing-library/react'
import { ReactNode } from 'react'

export * from '@testing-library/react'

const styles = fs.readFileSync(path.relative(process.cwd(), './lib/bundle.css'), 'utf8')

export function render(ui: ReactNode, options?: RenderOptions): RenderResult {
  const view = rtlRender(ui, options)

  const styleElement = document.createElement('style')
  styleElement.innerHTML = styles
  document.body.appendChild(styleElement)

  return view
}
