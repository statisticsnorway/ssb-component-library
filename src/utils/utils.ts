// utils.ts
const classNames = (...classes: (string | boolean | undefined)[]): string => classes.filter(Boolean).join(' ')

export default classNames
