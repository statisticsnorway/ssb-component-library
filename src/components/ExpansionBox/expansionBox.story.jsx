import React from 'react'
import ExpansionBox from './ExpansionBox'

const mockedHeader = 'This is a expansion box header'
const mockedText =
  // eslint-disable-next-line max-len
  'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.'

const mockedContent = (
  <div>
    <span>
      This is a normal paragraph text where the explanation of the fact is written. Here you can write a concise and
      descriptive text that helps the user understand the content. This is a normal paragraph text where the explanation
      of the fact is written. Here you can write a concise and descriptive text that helps the user understand the
      content.
    </span>

    <h3>Mulighet for tittel i boksen</h3>
    <ul>
      <li>Mulighet for punktliste</li>
      <li>Må være med</li>
      <li>Flere punkter</li>
    </ul>

    <p>
      <strong>Bold stiling</strong>: Det må også være rom for bold stiling av ord, avsnitt og setninger i boksen. This
      is a normal paragraph text where the explanation of the fact is written. Here you can write a concise and
      descriptive text that helps the user understand the content.
    </p>

    <span>
      This is a normal paragraph text <a href='https://www.ssb.no'>lenke i løpende tekst</a> where the explanation of
      the fact is written. Here you can write a concise and descriptive text that helps the user understand the content.
    </span>
  </div>
)

export default {
  title: 'ExpansionBox',
}

export const Default = () => (
  <div style={{ width: '750px' }}>
    <ExpansionBox header={mockedHeader} text={mockedContent} />
  </div>
)

export const DefaultOpen = () => (
  <div style={{ width: '750px' }}>
    <ExpansionBox header={mockedHeader} text={mockedContent} openByDefault />
  </div>
)

export const SneakPeek = () => (
  <div style={{ width: '740px', marginRight: '10px' }}>
    <ExpansionBox header='ExpansionBox with sneakpeek text' text={mockedText} sneakPeek />
  </div>
)

export const SneakpeekWithAiIcon = () => (
  <div style={{ width: '740px' }}>
    <ExpansionBox header='ExpansionBox with sneakpeek text and AI icon' text={mockedContent} aiIcon sneakPeek />
  </div>
)
