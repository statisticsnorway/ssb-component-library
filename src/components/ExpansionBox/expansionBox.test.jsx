import React from 'react'
import userEvent from '@testing-library/user-event'
import { screen, render } from '../../utils/test'
import ExpansionBox from './ExpansionBox'

const headerText = 'This is a ExpansionBox title'
const bodyText = 'This is paragraph text'

describe('ExpansionBox component', () => {
  test('Matches the snapshot', () => {
    const { asFragment } = render(<ExpansionBox header={headerText} text={bodyText} />)
    expect(asFragment()).toMatchSnapshot()
  })
  test('Starts closed', async () => {
    const { container } = render(<ExpansionBox header={headerText} text={bodyText} />)
    const element = container.querySelector('.content')
    expect(element).toHaveClass('closed')
  })
  test('Open on click when starting closed', async () => {
    const user = userEvent.setup()
    render(<ExpansionBox header={headerText} text={bodyText} />)
    const button = await screen.findByText(headerText)
    await user.click(button.parentElement)
    expect(await screen.findByText(bodyText)).toBeVisible()
  })
  test('set openByDefault', async () => {
    const text = 'ExpansionBox Text'
    render(<ExpansionBox header={headerText} text={bodyText} openByDefault />)
    expect(await screen.findByText(bodyText)).toBeVisible()
  })
})
