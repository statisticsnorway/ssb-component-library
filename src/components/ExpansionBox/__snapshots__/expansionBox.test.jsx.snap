// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpansionBox component Matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="ssb-expansion-box"
  >
    <button
      aria-expanded="false"
      class="header"
    >
      <span
        class="header-text"
      >
        This is a ExpansionBox title
      </span>
      <div
        class="icon-wrapper"
      >
        <svg
          class="expand-icon"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="6 9 12 15 18 9"
          />
        </svg>
      </div>
    </button>
    <div
      class="content closed"
    >
      This is paragraph text
    </div>
  </div>
</DocumentFragment>
`;
